apps/server/lib/lexical/server/state.ex

```diff
@@ -170,11 +170,20 @@ defmodule Lexical.Server.State do
     end
   end
 
-  def apply(%__MODULE__{} = state, %DidOpen{lsp: event}) do
-    %TextDocument.Item{text: text, uri: uri, version: version} =
-      text_document = event.text_document
-
-    case Document.Store.open(uri, text, version) do
+  def apply(%__MODULE__{} = state, %DidOpen{} = did_open) do
+    %DidOpen{
+      lsp: %DidOpen.LSP{
+        text_document:
+          %TextDocument.Item{
+            text: text,
+            uri: uri,
+            version: version,
+            language_id: language_id
+          } = text_document
+      }
+    } = did_open
```

Comment:
This is a lot more wordy than the previous version.
```suggestion
  def apply(%__MODULE__{} = state, %DidOpen{lsp: event} = did_open) do
    %TextDocument.Item{
            text: text,
            uri: uri,
            version: version,
            language_id: language_id
          } = event.text_document
```

Reply:
Also, this might be my issue, but we might not need to reach into the LSP fields. 
We can probably just do `did_open.text_document`

Reply:
I prefer de-structuring over direct de-referencing, as it keeps the body of the function more concise. I do like the de-structuring you suggested more though, applied it.

Reply:
A vs B:
```elixir
def apply(%__MODULE__{} = state, %DidOpen{lsp: event}) do
  %TextDocument.Item{
    text: text,
    uri: uri,
    version: version,
    language_id: language_id
  } = event.text_document
  
  case Document.Store.open(uri, text, language_id, version) do
    :ok ->
      Logger.info("opened #{uri}")
      {:ok, state}
  
  error ->
    Logger.error("Could not open #{uri} #{inspect(error)}")
    error
  end
end
```

```elixir
def apply(%__MODULE__{} = state, %DidOpen{lsp: event}) do
  text_document = event.text_document

  case Document.Store.open(
         text_document.uri,
         text_document.text,
         text_document.language_id,
         text_document.version
       ) do
    :ok ->
      Logger.info("opened #{text_document.uri}")
      {:ok, state}

    error ->
      Logger.error("Could not open #{text_document.uri} #{inspect(error)}")
      error
  end
end
```

I feel it's easier to grok at a glance with destructuring.

Reply:
My suggested change is not representative of suggestion B, so i don't understand the comment.

Reply:
Whoops, my bad. I think I see what you mean now, updated.